From f506a8875dc180b46b1434d0fade77834a64d8d4 Mon Sep 17 00:00:00 2001
From: Shengjiu Wang <shengjiu.wang@freescale.com>
Date: Fri, 17 Feb 2017 10:09:31 +0800
Subject: [PATCH 1529/5242] MLK-13992: ASoC: fsl_rpmsg_i2s: add flush
 workqueue

commit  0f337455089920d1ab8cfc2311ea4e19b03a19b1 from
https://source.codeaurora.org/external/imx/linux-imx.git

some cmd is sent by workqueue, others are sent by call send message
function directly, for workqueue may have delay, so there is occasion
that cmd is not sent in order.
Add flush_workqueue before the CLOSE and SUSPEND to make sure previous
cmd is finished in that time.

Signed-off-by: Shengjiu Wang <shengjiu.wang@freescale.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 sound/soc/fsl/fsl_rpmsg_i2s.c |    1 +
 sound/soc/fsl/imx-pcm-rpmsg.c |    2 ++
 2 files changed, 3 insertions(+)

diff --git a/sound/soc/fsl/fsl_rpmsg_i2s.c b/sound/soc/fsl/fsl_rpmsg_i2s.c
index 172c279..d8f8da8 100644
--- a/sound/soc/fsl/fsl_rpmsg_i2s.c
+++ b/sound/soc/fsl/fsl_rpmsg_i2s.c
@@ -196,6 +196,7 @@ static int fsl_rpmsg_i2s_suspend(struct device *dev)
 	struct i2s_rpmsg_s *rpmsg_tx;
 	struct i2s_rpmsg_s *rpmsg_rx;
 
+	flush_workqueue(i2s_info->rpmsg_wq);
 	rpmsg_tx = &i2s_info->send_msg[SNDRV_PCM_STREAM_PLAYBACK];
 	rpmsg_rx = &i2s_info->send_msg[SNDRV_PCM_STREAM_CAPTURE];
 
diff --git a/sound/soc/fsl/imx-pcm-rpmsg.c b/sound/soc/fsl/imx-pcm-rpmsg.c
index cbd5ea8..6c5f450 100644
--- a/sound/soc/fsl/imx-pcm-rpmsg.c
+++ b/sound/soc/fsl/imx-pcm-rpmsg.c
@@ -151,6 +151,8 @@ static int imx_rpmsg_pcm_close(struct snd_pcm_substream *substream)
 		rpmsg->header.cmd = I2S_TX_CLOSE;
 	else
 		rpmsg->header.cmd = I2S_RX_CLOSE;
+
+	flush_workqueue(i2s_info->rpmsg_wq);
 	i2s_info->send_message(rpmsg, i2s_info);
 
 	kfree(prtd);
-- 
1.7.9.5

